### PR TITLE
Importing mothur-specific datatypes

### PR DESCRIPTION
**Pull request by IyadKandalaft**
I am bringing in Jim Johnson's mothur datatypes but removed the QIIME and Tree datatypes from the datatypes_conf.xml and metagenomics.py.  I've also renamed metagenomics.py to mothur.py to create a namespace that will help avoid collisions.

TODO: Sublass datatypes that have little metadata.

### CODE REVIEW

**Comment by martenson** [comment]
This might want to target the galaxy repo similar to https://github.com/galaxyproject/galaxy/pull/1186

**Comment by lparsons** [comment]
Are any of these datatypes (e.g. otu) shared with Qiime? If so, perhaps we can keep them as more generic metagenomics datatypes? Or are the files actually of a different format?

**Comment by jj-umn** [comment]
I think it would be a good idea to prefix each of the mothur specific datatype extensions with: 'mothur.'  if they are to be merged into the main galaxy datatypes.

**Comment by IyadKandalaft** [comment]
From what I can tell, these are specific to mother but I'm not certain.
For example, OTU is no longer supported by QIIME or MEGAN5 but still
predominates in mothur.  I removed the non-mothur data types.

**Comment by IyadKandalaft** [comment]
Hey Jim.  Good to see you're still active.  Is a prefix required or just
the module name is enough?  I had started adding prefixes but I stopped
because I wasn't sure.

**Comment by jj-umn** [comment]
@IyadKandalaft 

The prefix would just help avoid name clashes, e.g.: 
<datatype extension="mothur.pair.dist" type="galaxy.datatypes.mothur:PairwiseDistanceMatrix" display_in_upload="true"/>

But of course all the data inputs and outputs will need to be modified as well.  I wish I had used the 'mothur.'  prefix initially, but I had no idea how datatypes would develop 5 years ago.

File: datatypes/mothur_datatypes/mothur.py [hunk 1]
```diff
@@ -0,0 +1,1172 @@
...
+        The first line has the number of sequences in the matrix.
+        The following lines have the sequence name in the first column plus a column for the distance to each sequence 
+        in the row order in which they appear in the matrix.
+               3
+               U68589  0.0000  0.3371  0.3610
+               U68590  0.3371  0.0000  0.3783
+               U68590  0.3371  0.0000  0.3783
+        """
+        try:
+            fh = open( filename )
```
Comment (yhoogstrate) [inline datatypes/mothur_datatypes/mothur.py]: `with open(filename,"r") as fh:` allows you to get rid of the `fh.close()` and the `finaly:`

File: datatypes/mothur_datatypes/mothur.py [hunk 2]
```diff
@@ -0,0 +1,1172 @@
...
+            line = fh.readline()
+            line = line.strip()
+            sequence_count = int(line)
+            col_cnt = seq_cnt + 1
+            while True:
+                line = fh.readline()
+                line = line.strip()
+                if not line:
+                    break #EOF
+                if line:
```
Comment (yhoogstrate) [inline datatypes/mothur_datatypes/mothur.py]: `if line and line[0] != '@'`:

File: datatypes/mothur_datatypes/mothur.py [hunk 3]
```diff
@@ -0,0 +1,1172 @@
...
+               U68590  0.3371  0.0000  0.3783
+               U68590  0.3371  0.0000  0.3783
+        """
+        try:
+            fh = open( filename )
+            count = 0
+            line = fh.readline()
+            line = line.strip()
+            sequence_count = int(line)
+            col_cnt = seq_cnt + 1
```
Comment (yhoogstrate) [inline datatypes/mothur_datatypes/mothur.py]: `s/seq_cnt/sequence_count/`

File: datatypes/mothur_datatypes/mothur.py [hunk 4]
```diff
@@ -0,0 +1,1172 @@
...
+            while True:
+                line = fh.readline()
+                line = line.strip()
+                if not line:
+                    break #EOF
+                if line:
+                    if line[0] != '@':
+                        linePieces = line.split('\t')
```
Comment (yhoogstrate) [inline datatypes/mothur_datatypes/mothur.py]: If the delimiter is `\t`, please give it in the example above (it currently uses 2 spaces)

**Comment by bgruening** [comment]
@lparsons as far as I know this are very specific mothur datatypes, hence the prefix idea. Shared, and more general metagenomic datatypes should stay more generic.
@IyadKandalaft do you want to submit this PR to Galaxy?

**Comment by IyadKandalaft** [comment]
I'll add the prefix mother prefix shortly. Do you think this belongs in the
galaxy repo? I thought that data types are typically maintained with the
tools. But I'm open to a PR with the galaxy repo.

**Comment by bgruening** [comment]
@IyadKandalaft yes the IUC and the Galaxy Core Team agreed to move datatypes into Core until we found a better solution. Thanks!

**Comment by bgruening** [comment]
@IyadKandalaft are you planning to work on this? I think this is the only piece that prevents us from putting the mothur wrappers as a first version into the TS.

Thanks!

File: datatypes/mothur_datatypes/mothur.py [hunk 5]
```diff
@@ -0,0 +1,1172 @@
...
+    def __init__(self, **kwd):
+        Text.__init__( self, **kwd )
+    def set_meta( self, dataset, overwrite = True, **kwd ):
+        if dataset.has_data():
+            label_names = set()
+            ncols = 0
+            data_lines = 0
+            comment_lines = 0
+            try:
+                fh = open( dataset.file_name )
```
Comment (blankenberg) [inline datatypes/mothur_datatypes/mothur.py]: ```with open```

File: datatypes/mothur_datatypes/mothur.py [hunk 6]
```diff
@@ -0,0 +1,1172 @@
...
+                        data_lines += 1
+                        ncols = max(ncols,len(fields))
+                        label_names.add(fields[0])
+                    else:
+                        comment_lines += 1
+                # Set the discovered metadata values for the dataset
+                dataset.metadata.data_lines = data_lines
+                dataset.metadata.columns = ncols
+                dataset.metadata.labels = []
+                dataset.metadata.labels += label_names
```
Comment (blankenberg) [inline datatypes/mothur_datatypes/mothur.py]: ```dataset.metadata.labels = list( label_names )```

File: datatypes/mothur_datatypes/mothur.py [hunk 7]
```diff
@@ -0,0 +1,1172 @@
...
+                dataset.metadata.labels.sort()
+            finally:
+                fh.close()
+
+    def sniff( self, filename ):
+        """
+        Determines whether the file is a otu (operational taxonomic unit) format
+        """
+        try:
+            fh = open( filename )
```
Comment (blankenberg) [inline datatypes/mothur_datatypes/mothur.py]: ```with open```

File: datatypes/mothur_datatypes/mothur.py [hunk 8]
```diff
@@ -0,0 +1,1172 @@
...
+    def init_meta( self, dataset, copy_from=None ):
+        Otu.init_meta( self, dataset, copy_from=copy_from )
+    def sniff( self, filename ):
+        """
+        Determines whether the file is a otu (operational taxonomic unit) format
+        label<TAB>count[<TAB>value(1..n)]
+        
+        """
+        try:
+            fh = open( filename )
```
Comment (blankenberg) [inline datatypes/mothur_datatypes/mothur.py]: ```with open```

File: datatypes/mothur_datatypes/mothur.py [hunk 9]
```diff
@@ -0,0 +1,1172 @@
...
+                    data_lines += 1
+                    fields = line.strip().split('\t')
+                    ncols = max(ncols,len(fields))
+                    label_names.add(fields[0])
+                    group_names.add(fields[1])
+                # Set the discovered metadata values for the dataset
+                dataset.metadata.data_lines = data_lines
+                dataset.metadata.columns = ncols
+                dataset.metadata.labels = []
+                dataset.metadata.labels += label_names
```
Comment (blankenberg) [inline datatypes/mothur_datatypes/mothur.py]: ```dataset.metadata.labels = list( label_names )```

File: datatypes/mothur_datatypes/mothur.py [hunk 10]
```diff
@@ -0,0 +1,1172 @@
...
+                dataset.metadata.labels.sort()
+                dataset.metadata.groups = []
+                dataset.metadata.groups += group_names
```
Comment (blankenberg) [inline datatypes/mothur_datatypes/mothur.py]: ```dataset.metadata.groups = list( group_names )```

File: datatypes/mothur_datatypes/mothur.py [hunk 11]
```diff
@@ -0,0 +1,1172 @@
...
+                fh.close()
+
+    def sniff( self, filename, vals_are_int=False):
+        """
+        Determines whether the file is a otu (operational taxonomic unit) Shared format
+        label<TAB>group<TAB>count[<TAB>value(1..n)]
+        The first line is column headings as of Mothur v 1.20
+        """
+        try:
+            fh = open( filename )
```
Comment (blankenberg) [inline datatypes/mothur_datatypes/mothur.py]: ```with open```

File: datatypes/mothur_datatypes/mothur.py [hunk 12]
```diff
@@ -0,0 +1,1172 @@
...
+        self.column_names = ['Map']
+
+    def sniff( self, filename ):
+        """
+        Determines whether the file is a secondary structure map format
+        A single column with an integer value which indicates the row that this row maps to.
+        check you make sure is structMap[10] = 380 then structMap[380] = 10.
+        """
+        try:
+            fh = open( filename )
```
Comment (blankenberg) [inline datatypes/mothur_datatypes/mothur.py]: ```with open```

File: datatypes/mothur_datatypes/mothur.py [hunk 13]
```diff
@@ -0,0 +1,1172 @@
...
+        """Initialize AlignCheck datatype"""
+
+    def sniff( self, filename ):
+        """
+        Determines whether the file is in Mothur align fasta format
+        Each sequence line must be the same length
+        """
+        
+        try:
+            fh = open( filename )
```
Comment (blankenberg) [inline datatypes/mothur_datatypes/mothur.py]: ```with open```

File: datatypes/mothur_datatypes/mothur.py [hunk 14]
```diff
@@ -0,0 +1,1172 @@
...
+        Tabular.__init__( self, **kwd )
+        self.column_names = ['name','pound','dash','plus','equal','loop','tilde','total']
+        self.column_types = ['str','int','int','int','int','int','int','int']
+        self.comment_lines = 1
+
+    def set_meta( self, dataset, overwrite = True, **kwd ):
+        # Tabular.set_meta( self, dataset, overwrite = overwrite, first_line_is_header = True, skip = 1 )
+        data_lines = 0
+        if dataset.has_data():
+            dataset_fh = open( dataset.file_name )
```
Comment (blankenberg) [inline datatypes/mothur_datatypes/mothur.py]: ```with open```

File: datatypes/mothur_datatypes/ref_to_seq_taxonomy_converter.xml [hunk 15]
```diff
@@ -0,0 +1,12 @@
+<tool id="CONVERTER_ref_to_seq_taxomony" name="Convert Ref taxonomy to Seq Taxonomy" version="1.0.0">
+  <description>converts 2 or 3 column sequence taxonomy file to a 2 column mothur taxonomy_outline format</description>
+  <command interpreter="python">ref_to_seq_taxonomy_converter.py $input $output</command>
```
Comment (blankenberg) [inline datatypes/mothur_datatypes/ref_to_seq_taxonomy_converter.xml]: quote ```$input``` ```$output```

**Comment by blankenberg** [comment]
@IyadKandalaft @bgruening Is there a plan?

**Comment by bgruening** [comment]
I hope I could convince @oxyko to work on this and move it over into Galaxy. Any help here is appreciated. As soon as we have the datatypes in we can merge https://github.com/galaxyproject/tools-iuc/pull/429, which is not perfect, but a huge step forward.

**Comment by shiltemann** [comment]
It would make my life easier with writing tests for the tools if the datatypes are in, so I'll work on this and move it into Galaxy soon :)

File: datatypes/mothur_datatypes/mothur.py [hunk 16]
```diff
@@ -0,0 +1,1172 @@
...
+                    return True
+                else:
+                    return False
+        except:
+            pass
+        finally:
+            fh.close()
+        return False
+
+class AlignCheck(Tabular):
```
Comment (shiltemann) [inline datatypes/mothur_datatypes/mothur.py]: This seems to be a redefinition of unused AlignCheck on line 315?

Comment (bgruening) [inline datatypes/mothur_datatypes/mothur.py]: Seems so, wired ...

**Comment by oxyko** [comment]
Looks good, Saskia! Thanks for taking this on.

**Comment by bgruening** [comment]
@IyadKandalaft can we close this PR.
@shiltemann integrated many parts of this PR in: https://github.com/galaxyproject/galaxy/issues/2038

**Comment by IyadKandalaft** [comment]
sure go ahead.

**Comment by bgruening** [comment]
Thanks all! Let's rock the other branch.

**Pull request closed without merging**